### PR TITLE
backendaddresspool:  Ensure unique addresses; Simplify

### PR DESCRIPTION
**Pull request by draychev**
As I'm working to understand what it would take to add multiple namespace support to the k8s ingress controller, I decided to dig into the creation of backend address pools.

To make it easier to understand - split the code in small functions.
In addition to that - `getPorts` ensures that we get unique ports and `getAddressesForSubset` returns unique addresses. (Was causing issues prior to this change)

The changes here are still tested by `appgw_test.go` --> https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/pkg/appgw/appgw_test.go#L237
To ensure these tests continue to pass I added the `byIPFQDN ` sorter.

### CODE REVIEW

File: pkg/appgw/backendaddresspools.go [hunk 1]
```diff
@@ -7,82 +7,121 @@ package appgw
...
 
-					addressPools[*addressPool.Name] = addressPool
-				}
+func (builder *appGwConfigBuilder) getBackendAddressPool(backendID backendIdentifier, serviceBackendPair serviceBackendPortPair, defaultPool *n.ApplicationGatewayBackendAddressPool, addressPools *map[string]*n.ApplicationGatewayBackendAddressPool) *n.ApplicationGatewayBackendAddressPool {
+	endpoints := builder.k8sContext.GetEndpointsByService(backendID.serviceKey())
+	if endpoints == nil {
+		logLine := fmt.Sprintf("Unable to get endpoints for service key [%s]", backendID.serviceKey())
+		builder.recorder.Event(backendID.Ingress, v1.EventTypeWarning, "EndpointsEmpty", logLine)
+		glog.Warning(logLine)
+		return defaultPool
```
Comment (akshaysngupta) [inline pkg/appgw/backendaddresspools.go]: `return nil` here to be more explicit that this function did nothing ?

File: pkg/appgw/backendaddresspools.go [hunk 2]
```diff
@@ -7,82 +7,121 @@ package appgw
...
+		Name: &poolName,
+		ApplicationGatewayBackendAddressPoolPropertiesFormat: &n.ApplicationGatewayBackendAddressPoolPropertiesFormat{
+			BackendAddresses: getAddressesForSubset(subset),
+		},
 	}
+}
 
-	builder.appGwConfig.BackendAddressPools = &backendPools
+func getAddressesForSubset(subset v1.EndpointSubset) *[]n.ApplicationGatewayBackendAddress {
+	// We make separate maps for IP and FQDN to ensure uniqueness within the 2 groups
```
Comment (akshaysngupta) [inline pkg/appgw/backendaddresspools.go]: Why do we need to two separate maps for IP and FQDN ?

File: pkg/appgw/backendaddresspools.go [hunk 3]
```diff
@@ -7,82 +7,121 @@ package appgw
...
+	}
 
-	return builder, nil
+	for ip := range ips {
+		addrSet[n.ApplicationGatewayBackendAddress{IPAddress: to.StringPtr(ip)}] = nil
+	}
+	for fqdn := range fqdns {
+		addrSet[n.ApplicationGatewayBackendAddress{IPAddress: to.StringPtr(fqdn)}] = nil
+	}
+	return getBackendAddressMapKeys(&addrSet)
```
Comment (akshaysngupta) [inline pkg/appgw/backendaddresspools.go]: can we move the `getBackendAddressMapKeys` below this function ?
I like a flow where a function is used first to establish some context and then defined in the next section.

File: pkg/appgw/backendaddresspools.go [hunk 4]
```diff
@@ -7,82 +7,121 @@ package appgw
...
+			poolName := generateAddressPoolName(backendID.serviceFullName(), backendID.Backend.ServicePort.String(), serviceBackendPair.BackendPort)
+			// The same service might be referenced in multiple ingress resources, this might result in multiple `serviceBackendPairMap` having the same service key but different
+			// ingress resource. Thus, while generating the backend address pool, we should make sure that we are generating unique backend address pools.
+			pool, ok := (*addressPools)[poolName]
+			if !ok {
+				pool = newPool(poolName, subset)
+				(*addressPools)[poolName] = pool
 			}
+			return pool
 		}
```
Comment (akshaysngupta) [inline pkg/appgw/backendaddresspools.go]: If we are not able to find the `serviceBackendPair.BackendPort`, should we raise an event that some subnet in endpoint don't have the referenced `port` ?

File: pkg/appgw/backendaddresspools.go [hunk 5]
```diff
@@ -7,82 +7,121 @@ package appgw
...
-			glog.Warning(logLine)
-			builder.backendPoolMap[backendID] = &emptyPool
-			continue
+func (builder *appGwConfigBuilder) BackendAddressPools(ingressList []*v1beta1.Ingress) (ConfigBuilder, error) {
+	defaultPool := defaultBackendAddressPool()
+	addressPools := map[string]*n.ApplicationGatewayBackendAddressPool{
+		*defaultPool.Name: &defaultPool,
+	}
+	for backendID, serviceBackendPair := range builder.getServiceBackendPairMap() {
+		if pool := builder.getBackendAddressPool(backendID, serviceBackendPair, &defaultPool, &addressPools); pool != nil {
```
Comment (akshaysngupta) [inline pkg/appgw/backendaddresspools.go]: How about? No side-effects. :)
```go
    builder.backendPoolMap[backendID] = defaultPool
    if pool := builder.getBackendAddressPool(backendID, serviceBackendPair); pool != nil {
        addressPools[pool.Name] = pool
        builder.backendPoolMap[backendID] = pool
    }
```

File: pkg/appgw/backendaddresspools.go (hunk 1 above)
Comment (draychev) [reply pkg/appgw/backendaddresspools.go]: Reading the code how it was before the refactor - we want to return the `defaultPool` (what was formerly `emptyPool`).  Here is where this is coming from:  https://github.com/Azure/application-gateway-kubernetes-ingress/blob/a495cc5d2674c1bf18b46574e3e6631824a776a7/pkg/appgw/backendaddresspools.go#L23-L29

File: pkg/tests/types.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package tests
+
+import (
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// MockEventRecorder mocks EventRecorder, which knows how to record events on behalf of an EventSource.
+type MockEventRecorder struct{}
```
Comment (akshaysngupta) [inline pkg/tests/types.go]: How about using `record.NewFakeRecorder` ?
https://github.com/Azure/application-gateway-kubernetes-ingress/blob/c9cd1446aecf5d827ca9f967d51c64fd43aa180d/pkg/appgw/appgw_test.go#L401

File: pkg/appgw/backendaddresspools.go [hunk 7]
```diff
@@ -7,82 +7,126 @@ package appgw
...
-			builder.recorder.Event(backendID.Ingress, v1.EventTypeWarning, "EndpointsEmpty", logLine)
-			glog.Warning(logLine)
-			builder.backendPoolMap[backendID] = &emptyPool
-			continue
+func (builder *appGwConfigBuilder) BackendAddressPools(ingressList []*v1beta1.Ingress) (ConfigBuilder, error) {
+	defaultPool := defaultBackendAddressPool()
+	addressPools := map[string]*n.ApplicationGatewayBackendAddressPool{
+		*defaultPool.Name: defaultPool,
+	}
+	for backendID, serviceBackendPair := range builder.getServiceBackendPairMap() {
```
Comment (akshaysngupta) [inline pkg/appgw/backendaddresspools.go]: `builder.backendPoolMap[backendID] = defaultPool`

File: pkg/appgw/backendaddresspools.go [hunk 8]
```diff
@@ -7,82 +7,126 @@ package appgw
...
-				// ingress resource. Thus, while generating the backend address pool, we should make sure that we are generating unique backend address pools.
-				addressPool, ok := addressPools[addressPoolName]
+func (builder *appGwConfigBuilder) getBackendAddressPool(backendID backendIdentifier, serviceBackendPair serviceBackendPortPair, addressPools map[string]*n.ApplicationGatewayBackendAddressPool) *n.ApplicationGatewayBackendAddressPool {
+	endpoints := builder.k8sContext.GetEndpointsByService(backendID.serviceKey())
+	if endpoints == nil {
+		logLine := fmt.Sprintf("Unable to get endpoints for service key [%s]", backendID.serviceKey())
+		// TODO(draychev): Move "reason" into an enum
+		builder.recorder.Event(backendID.Ingress, v1.EventTypeWarning, "EndpointsEmpty", logLine)
+		glog.Warning(logLine)
+		return defaultBackendAddressPool()
```
Comment (akshaysngupta) [inline pkg/appgw/backendaddresspools.go]: We can combine `return nil` with `builder.backendPoolMap[backendID] = defaultPool`?

File: pkg/tests/types.go (hunk 6 above)
Comment (draychev) [reply pkg/tests/types.go]: This is great - always happy to reuse great stuff!

File: pkg/appgw/backendaddresspools.go (hunk 8 above)
Comment (draychev) [reply pkg/appgw/backendaddresspools.go]: I like this!

**Pull request merged**